Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMessagePortCocoa.mm

```diff
@@ -0,0 +1,147 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#import "config.h"
+#import "WebExtensionMessagePort.h"
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "CocoaHelpers.h"
+#import "WebExtensionContext.h"
+#import "_WKWebExtensionMessagePortInternal.h"
+#import <wtf/BlockPtr.h>
+
+namespace WebKit {
+
+NSError *toAPI(WebExtensionMessagePort::Error error)
+{
+    if (!error)
+        return nil;
+
+    _WKWebExtensionMessagePortError errorCode;
+    NSString *message;
+
+    switch (error.value().first) {
+    case WebKit::WebExtensionMessagePort::ErrorType::Unknown:
+        errorCode = _WKWebExtensionMessagePortErrorUnknown;
+        message = (NSString *)error.value().second.value_or("An unknown error occurred."_s);
+        break;
+
+    case WebKit::WebExtensionMessagePort::ErrorType::NotConnected:
+        errorCode = _WKWebExtensionMessagePortErrorNotConnected;
+        message = (NSString *)error.value().second.value_or("Message port is not connected and cannot send messages."_s);
+        break;
+    }
+
+    return [NSError errorWithDomain:_WKWebExtensionMessagePortErrorDomain code:errorCode userInfo:@{ NSDebugDescriptionErrorKey: message }];
+}
+
+WebExtensionMessagePort::Error toWebExtensionMessagePortError(NSError *error)
+{
+    if (!error)
+        return std::nullopt;
+    return { { WebExtensionMessagePort::ErrorType::Unknown, error.localizedDescription } };
+}
+
+WebExtensionMessagePort::WebExtensionMessagePort(WebExtensionContext& extensionContext, String applicationIdentifier, WebExtensionPortChannelIdentifier channelIdentifier)
+    : m_extensionContext(extensionContext)
+    , m_applicationIdentifier(applicationIdentifier)
+    , m_channelIdentifier(channelIdentifier)
+{
+}
+
+bool WebExtensionMessagePort::operator==(const WebExtensionMessagePort& other) const
+{
+    return this == &other || (m_extensionContext == other.m_extensionContext && m_applicationIdentifier == other.m_applicationIdentifier && m_channelIdentifier == other.m_channelIdentifier);
+}
+
+WebExtensionContext* WebExtensionMessagePort::extensionContext() const
+{
+    return m_extensionContext.get();
+}
+
+bool WebExtensionMessagePort::isDisconnected() const
+{
+    return !m_extensionContext;
+}
+
+void WebExtensionMessagePort::disconnect(Error error)
+{
+    if (isDisconnected())
+        return;
+
+    m_extensionContext->portDisconnect(WebExtensionContentWorldType::Native, WebExtensionContentWorldType::Main, m_channelIdentifier);
+
+    remove();
+}
+
+void WebExtensionMessagePort::reportDisconnection(Error error)
+{
+    ASSERT(!isDisconnected());
```

Comment:
hmm i'm a bit confused on why the port wouldn't already be disconnected if we're reporting that it's been disconnected

Reply:
This is the internal method that triggers the disconnect. It should only be called once, so it ASSERTs. There is a public disconnect() that is exposed to JS, and it returns early if already disconnected, since that is the behavior in other browsers and it is safe to call multiple times.